### ISSUE TITLE
[AUC] result of tf.metrics.auc doesnot match with sklearn's

### ISSUE THREAD

**Issue by maybeluo** [tensorflow/tensorflow#14834]
My tensorflow version is ('v1.3.0-rc1-4263-gc81acfb', '1.4.0-rc1'), and the system is Rehat with gcc version 4.8.5 20150623 (Red Hat 4.8.5-16). I run the program use CPU only.

I wrote a NN  use tensorflow for binary classification. The output layer has only ONE unit and the loss function is `tf.nn.sigmoid_cross_entropy_with_logits`. I create the an `auc_op` in the following way:
 
```python
net = input_layer(features,) # get dense input layer from features
for layer_id in xrange(1, num_layer):
    net  = tf.add(tf.matmul(net, self._weights[layer_id]), self._bias[layer_id])
    if layer_id < num_layer - 1: # output layer without activation function to get `wx + b`
        net =tf.nn.relu(net)
logits = net
labels = tf.expand_dims(tf.cast(tf.convert_to_tensor(labels), dtype = tf.float32), axis = -1)
auc_op = tf.metrics.auc(labels = labels, predictions = tf.sigmoid(logits), num_thresholds = 102400)
```

I run the `auc_op` like this:
```python
for step in xrange(1, self._max_steps + 1):
    auc = self._sess.run(auc_op)
```

I also keep all the logits and labels in each step and concatenate them, then call sklearn like this:
```python
from sklearn.metrics import roc_auc_score
roc_auc_score(labels, sigmoid(logits))
```

Are there anything wrong in the way I use tf.metrics.auc? When I run the `auc_op`, it returns a tuple with two values and I don't which one is the correct auc. But both of them are not equal with sklearn's. 
I once wrote an program to calculate auc and it was exactly the same with sklearn's even in 1M data, thus I tend to think sklearn's result is the ground truth.

**Comment by facaiy** [comment]
The API of `tf.metrics.auc` says that it is an **approximate** AUC via a Riemann sum. So I expect that the results will be a slightly different:

```python
import numpy as np
from sklearn.metrics import roc_auc_score
y_true = np.array([0, 0, 1, 1])
y_scores = np.array([0.1, 0.4, 0.35, 0.8])
print("sklearn auc: {}".format(roc_auc_score(y_true, y_scores)))

import tensorflow as tf
auc, update_op = tf.metrics.auc(y_true, y_scores)
with tf.Session() as sess:
    sess.run(tf.local_variables_initializer())
    print("tf auc: {}".format(sess.run([auc, update_op])))
```

```bash
~/Downloads ❯❯❯ python test.py
sklearn auc: 0.75
tf auc: [0.74999976, 0.74999976]
```

**Comment by maybeluo** [comment]
Thanks for your help. Two more questions:
1. On my dataset, the result from tensorflow and sklearn are quite different, the gap is about 0.5%-1%. I think this is a gap we cannot ignore.
2. I'm running mini-batch SGD, take 100 steps for example. For each step we run `tf.metrics.auc`, we will get two auc, one from `auc` and another is `update_op` in your code. They are NOT equal. Which one is the answer? Is `auc`  the result on all the previous data, or just on the current batch data?

**Comment by facaiy** [comment]
@maybeluo I agree that 1% gap is unsatisfied. It could be useful if you can provide some data for test (or benchmark report) , for example, a numpy array or csv file.

**Comment by maybeluo** [comment]
I'll do more experiments and look at the approximation method. Thanks.

**Comment by thkfly** [comment]
I got auc 0.78 on tensorflow,while I got 0.89 with sklearn

**Comment by Nicolabo** [comment]
Was the problem solved somehow? In my case i got AUC 0.82 on tensorflow and 0.90 in `sklearn.metrics.auc_score`.

**Comment by jerrymomo10** [comment]
got the same problem

**Comment by bis-carbon** [comment]
Another issue with ``` tf.metrics.auc ``` is, it doesn't give the same result when using one hot encoding and [0, 1] encoding. For example, ```target = [1, 1, 0, 1, 0] and one hot encoding of target is targ_one_hot=[[0, 1], [0, 1], [1, 0], [0, 1], [1, 0]] ``` if we these different representation   ``` tf.metrics.auc ``` gives different results where as ```sklearn.metrics.auc_score``` results are similar.

**Comment by berzjackson** [comment]
Just ran into similar problems and found this thread. If anyone else find here as well, hope this can help.

If your model's predictions are generally very small, say something less than 0.001, try increasing the **num_thresholds** argument in your auc call. The default value is set to 200, which in my case is too small and cause huge approximation error(as high as 0.2).

This is actually clearly stated in the [doc](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/AUC) as a caveat, but I do this auc metric in a future version should have better default behaviors or warning message for such cases.

> The num_thresholds variable controls the degree of discretization with larger numbers of thresholds more closely approximating the true AUC. The quality of the approximation may vary dramatically depending on num_thresholds. The thresholds parameter can be used to manually specify thresholds which split the predictions more evenly.

> For best results, predictions should be distributed approximately uniformly in the range [0, 1] and not peaked around 0 or 1. The quality of the AUC approximation may be poor if this is not the case. Setting summation_method to 'minoring' or 'majoring' can help quantify the error in the approximation by providing lower or upper bound estimate of the AUC.

**Comment by fritshermans** [comment]
To add to @berzjackson response; the summation method of sklearn seems to be different as well. I got almost the same results by using the following settings: 

`tf.keras.metrics.AUC(curve='PR', summation_method='majoring', thresholds=np.arange(0,1+0.000001,0.000001))`
